src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1479,16 +1531,61 @@ HasUnsupportedReferenceTableJoin(PlannerRestrictionContext *plannerRestrictionCo
 
 
 /*
- * RelationInfoHasReferenceTable check whether the relationInfo has reference table.
+ * RelationInfoContainsReferenceTable check whether the relationInfo
+ * has reference table.
+ *
  * Since relation ids of relationInfo indexes to the range table entry list of
  * planner info, planner info is also passed.
+ *
+ * Join relation infos are ignored since we're already going to process each of them
+ * seperately.
  */
 static bool
-RelationInfoHasReferenceTable(PlannerInfo *plannerInfo, RelOptInfo *relationInfo)
+RelationInfoContainsReferenceTable(PlannerInfo *plannerInfo, RelOptInfo *relationInfo)
 {
 	Relids relids = bms_copy(relationInfo->relids);
 	int relationId = -1;
 
+	/*
+	 * We shouldn't recursively go down for joins since we're already
+	 * going to process each join seperately. Otherwise we'd restrict
+	 * the coverage. See the below sketch where (h) denotes a hash
+	 * distributed relation, (r) denotes a reference table, (L) denotes
+	 * LEFT JOIN and (I) denotes INNER JOIN. If we're to recurse into
+	 * the inner join, we'd be preventing to push down the following
+	 * join tree, which is actually safe to push down.
+	 *
+	 *                       (L)
+	 *                      /  \
+	 *                   (I)     h
+	 *                  /  \
+	 *                r      h
+	 */
+	if (relationInfo->reloptkind == RELOPT_JOINREL)
+	{
+		return false;
+	}
+
+	/*
+	 * Note that we treat the same query where relations appear in subqueries
+	 * differently. (i.e., use SELECT * FROM r; instead of r)
+	 *
+	 * In that case, to relax some restrictions, we do the following optimization:
+	 * If the subplan (i.e., plannerInfo corresponding to the subquery) contains any
+	 * joins, we skip reference table checks keeping in mind that the join is already
+	 * going to be processed seperately. This optimization should suffice for many
```

Comment:
I feel like these `if` statements don't really belong in this function. The caller should decide whether or not to check a `RelationInfo` for reference tables and that's where the reasons why should be documented, otherwise code gets obscure.

Maybe in `HasUnsupportedReferenceTableJoin` we can do something like:
```
if (IsJoinRelationInfo(outerrel))
{
    /* description of why we skip this */
    continue;
}

if (IsSimpleSubquery(outerrel))
{
    /* description of why we skip this */
    continue;
}

if (joinType == JOIN_SEMI || joinType == JOIN_ANTI || joinType == JOIN_LEFT)
{
    if (!RelationInfoResultIsSharded(plannerInfo, outerrel))
    {
        return true;
    }
}
````

Reply:
These all make sense, I started in that way but somehow ended up with the current look. The trickier part was `FULL JOIN` where the code becomes a bit complicated.

Anyway, I'll try to shape it better.